azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,96 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This is a placeholder policy engine class to test that the regorus
+# dependency is correctly installed.
+# pylint: disable=too-few-public-methods
+
+import pwd
+import os
+import sys
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.version import get_distro
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common import conf
+
+def log_policy_info(formatted_string, op=WALAEventOperation.Policy, send_event=True):
+    """Log information to console and telemetry."""
+    logger.info("[Policy] " + formatted_string)
+    if send_event:
+        add_event(op=op, message=formatted_string)
+
+
+class PolicyEngine:
+    """Base class for policy engine"""
+    _instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance is None:
+            cls._instance = cls()
+        return cls._instance
+    
+    @staticmethod
+    def _is_policy_supported():
+        distro_info = get_distro()
+        distro_name = distro_info[0]
+        try:
+            distro_version = FlexibleVersion(distro_info[1])
+        except ValueError:
+            return False
+        return distro_name.lower() == 'ubuntu' and distro_version.major >= 16
+
+    def __init__(self):
+        self._initialized = False
+        self._policy_supported = False
+        self._extension_policy_enabled = False
+        self._engine = None
+
+    def initialize(self):
+        try:
+            if self._initialized:
+                return
+            self._policy_supported = self._is_policy_supported()
+
+            if not conf.get_extension_policy_enabled():
+                log_policy_info("Extension policy enforcement is disabled")
+                return
+
+            if not self._policy_supported:
+                log_policy_info("Policy enforcement is not supported on {0}".format(get_distro()))
+                return
+
+            # TO DO - import code is for e2e testing, remove once the binary has been published to GA package
+            user = next((u.pw_name for u in pwd.getpwall() if u.pw_uid == 1000), None)
+            regorus_dir = os.path.join(("/home/" + str(user)), "lib/tests_e2e/tests/executables")
+            sys.path.append(regorus_dir)
+            import regorus  # pylint: disable=import-outside-toplevel
+
+            self._engine = regorus.Engine()
+            self._extension_policy_enabled = True
+        except Exception as exception:
+            log_policy_info("Error initializing policy engine: {0}".format(exception))
+        finally:
+            log_policy_info("Policy enforcement enabled: {0}".format(self._extension_policy_enabled))
```

Comment:
I thought finally will be executed regardless whether exception is thrown or not. If my understanding is correct, when exception is thrown, the initialized flag should not be set to true. also due to its usage where the exception is not handled, you need to make sure all exceptions are caught and handled.

Reply:
Initialized should be set to true whether or not an exception is thrown, because the top-level singleton will still be initialized (I could rename this class to "PolicyConfigurator" to make the distinction a bit clearer). But if an exception is thrown, _extension_policy_enabled will be set to false and the underlying Regorus engine won't be initialized. The top-level class will still track if policy is enabled, and if an exception is ever thrown, policy will be globally disabled. Does that make sense?

Reply:
Doesn't quite make sense to me, sorry. The object may be half-initialized yet be marked as initialized, as in the example I posted above when there is an exception importing the regorus module.